Editor/Windows/AssetPaletteWindowEntryPanel.cs

```diff
@@ -230,72 +231,157 @@ private void DrawEntriesPanel()
 
             float containerWidth = Mathf.Floor(EditorGUIUtility.currentViewWidth) - FolderPanelWidth;
 
-            const float padding = 2;
-            int entrySize = Mathf.RoundToInt(Mathf.Lerp(EntrySizeMin, EntrySizeMax, ZoomLevel));
+            GUILayout.Space(Padding);
+
+            bool didClickASpecificEntry;
+            if (ZoomLevel == 0)
+            {
+                DrawListEntries(containerWidth, out didClickASpecificEntry);
+            }
+            else
+            {
+                DrawGridEntries(containerWidth, out didClickASpecificEntry);
+            }
+
+            GUILayout.Space(Padding);
+
+            // If you didn't click an entry and weren't pressing SHIFT, clear the selection.
+            if (Event.current.type == EventType.MouseDown && Event.current.button == 0 && !didClickASpecificEntry &&
+                !Event.current.shift)
+            {
+                StopAllRenames();
+
+                ClearEntrySelection();
+                Repaint();
+            }
+
+            EditorGUILayout.EndScrollView();
+        }
+
+        private void DrawListEntries(float containerWidth, out bool didClickASpecificEntry)
+        {
+            didClickASpecificEntry = false;
+
+            for (int i = 0; i < GetEntryCount(); i++)
+            {
+                PurgeInvalidEntries(i);
+                if (i >= GetEntryCount())
+                    break;
+
+                DrawListEntry(i, containerWidth, ref didClickASpecificEntry);
+            }
+        }
+
+        private void DrawListEntry(int index, float containerWidth, ref bool didClickASpecificEntry)
+        {
+            PaletteEntry entry = GetEntry(index);
+            Rect rect = GUILayoutUtility.GetRect(0, 0,
+                GUILayout.Width(containerWidth), GUILayout.Height(EditorGUIUtility.singleLineHeight));
+
+            HandleEntrySelection(index, rect, entry, ref didClickASpecificEntry);
+
+            Rect entryContentsRect = rect;
+            SerializedProperty entryProperty = SelectedFolderEntriesSerializedProperty.GetArrayElementAtIndex(index);
 
+            if (entry.IsRenaming)
+            {
+                string renameControlId = GetRenameControlId(entryProperty);
```

Comment:
Hmm, on second thought, can we extract this renaming part to its own method?

`DrawGridEntry` does the exact same thing, just at a different Rect.

Reply:
Sure thing!